src/Concretize.hs

```diff
@@ -677,26 +678,29 @@ concretizeDefinition allowAmbiguity typeEnv globalEnv visitedDefinitions definit
 -- For all other functions, the name must match exactly, and in all cases, the signature must match.
 allImplementations :: TypeEnv -> Env -> String -> Ty -> [(Env, Binder)]
 allImplementations typeEnv env functionName functionType =
-  filter (predicate . xobjTy . binderXObj . snd) foundBindings
+  (filter (predicate . xobjTy . binderXObj . snd) foundBindings)
   where
     predicate (Just t) =
       --trace ("areUnifiable? " ++ show functionType ++ " == " ++ show t ++ " " ++ show (areUnifiable functionType t)) $
       areUnifiable functionType t
     predicate Nothing = error "allfunctionswithnameandsignature"
-    foundBindings = case lookupBinder (SymPath [] functionName) (getTypeEnv typeEnv) of
+    foundBindings = case getTypeBinder typeEnv functionName of
       -- this function is an interface; lookup implementations
-      Just (Binder _ (XObj (Lst (XObj (Interface _ paths) _ _ : _)) _ _)) ->
+      Right (Binder _ (XObj (Lst (XObj (Interface _ paths) _ _ : _)) _ _)) ->
         -- N.B./TODO: There are functions designed for this
         -- scenario--e.g. lookupImplementations, but they cause
         -- either entirely unacceptable behavior (not finding
         -- implementations, or hangs). We should be able to use
         -- those here instead of looking up all interface paths
         -- directly, but for now we are stuck with this.
-        case sequence $ map (\p -> lookupInEnv p env) (paths ++ [(SymPath [] functionName)]) of
-          Just found -> found
-          Nothing -> (multiLookupEverywhere functionName env)
+        case sequence $ map (\p -> searchValue env p) (paths ++ [(SymPath [] functionName)]) of
+          Right found -> found
+          Left _ ->
+            case findPoly env functionName functionType of
+              Right r -> [r]
+              Left _ -> (lookupEverywhere env functionName)
       -- just a regular function; look for it
-      _ -> (multiLookupEverywhere functionName env)
+      _ -> fromRight [] ((fmap (: []) (Env.getValue env functionName)) <> pure (lookupEverywhere env functionName))
```

Comment:
The old function was just `multiLookupEverywhere functionName env` – is there a chance that this impenetrable line can be replaced by such a function again? Or is that "feature" only used here?

Reply:
If we only called `(lookupEverywhere env functionName))` it would amount to the same thing but lead to cases where we process a whole chain of envs when all we really need is the one env (when the function is a direct child)